tools/parsergen/tests/test_scoped_append_backend_fences.py

```
@@ -0,0 +1,107 @@
+            ),
+        ),
+    )
+    return source, changed
+
+
+@pytest.mark.parametrize("variant", (_source_only_variant, _ir_only_variant))
+def test_canonical_bsl_rejects_scoped_source_or_live_ir_before_entrypoints(variant) -> None:
+    parsed, _, _, parser_ir = _parts()
+    source, changed_ir = variant(parsed.source_grammar, parser_ir)
+
+    with pytest.raises(ValueError, match="scoped append.*canonical BSL"):
```

Ruff (0.16.3)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)